plugins/by-name/lint/default.nix

```diff
@@ -225,43 +159,70 @@ in
         };
       };
-  };
 
-  config = mkIf cfg.enable {
-    extraPlugins = [ cfg.package ];
+    linters =
+      defaultNullOpts.mkAttrsOf mkLinterOpts
+        {
+          phpcs.args = [
+            "-q"
+            "--report=json"
+            "-"
+          ];
+        }
+        ''
+          Configure the linters you want to run.
+          You can also add custom linters here.
+        '';
+
+    linters_by_ft =
+      defaultNullOpts.mkAttrsOf (types.listOf types.str)
+        {
+          text = [ "vale" ];
+          json = [ "jsonlint" ];
+          markdown = [ "vale" ];
+          rst = [ "vale" ];
+          ruby = [ "ruby" ];
+          janet = [ "janet" ];
+          inko = [ "inko" ];
+          clojure = [ "clj-kondo" ];
+          dockerfile = [ "hadolint" ];
+          terraform = [ "tflint" ];
+        }
+        ''
+          Configure the linters you want to run per file type.
+        '';
+  };
 
-    extraConfigLua =
+  imports = [
+    # Deprecations
+    (mkRenamedOptionModule [ "plugins" "lint" "lintersByFt" ] [ "plugins" "lint" "linters_by_ft" ])
+    (mkRemovedOptionModule [
+      "plugins"
+      "lint"
+      "customLinters"
+    ] "plugins.lint.customLinters option was removed, use linters instead.")
+  ];
+
+  extraConfig = cfg: {
+    autoCmd = lib.optionals (cfg.autoCmd != null) [ cfg.autoCmd ];
+    plugins.lint.luaConfig.content =
       ''
-        __lint = require('lint')
-        __lint.linters_by_ft = ${helpers.toLuaObject cfg.lintersByFt}
+        local lint = require('lint')
       ''
-      + (optionalString (cfg.linters != { }) (
-        concatLines (
-          flatten (
-            mapAttrsToList (
-              linter: linterConfig:
-              mapAttrsToList (
-                propName: propValue:
-                optionalString (
-                  propValue != null
-                ) ''__lint.linters["${linter}"]["${propName}"] = ${helpers.toLuaObject propValue}''
-              ) linterConfig
-            ) cfg.linters
-          )
-        )
-      ))
-      + (optionalString (cfg.customLinters != { }) (
-        concatLines (
-          mapAttrsToList (
-            customLinter: linterConfig:
+      + (lib.optionalString (cfg.linters_by_ft != null) ''
+        lint.linters_by_ft = ${toLuaObject cfg.linters_by_ft}
+      '')
+      + (lib.optionalString (cfg.linters != null) (
+        lib.concatLines (
+          lib.mapAttrsToList (
+            linter: linterConfig:
             let
-              linterConfig' = if isString linterConfig then helpers.mkRaw linterConfig else linterConfig;
+              linterConfig' =
+                if builtins.isString linterConfig then lib.nixvim.mkRaw linterConfig else linterConfig;
             in
-            ''__lint.linters["${customLinter}"] = ${helpers.toLuaObject linterConfig'}''
-          ) cfg.customLinters
+            ''lint.linters.${linter} = ${toLuaObject linterConfig'}''
+          ) cfg.linters
```

Comment:
>Do all this plugin's options take the form `require('lint')[optname] = optval?` And there is no `.setup()` or similar function?

Yes, this plugin doesn't have `.setup()`. Which I quite like, to be honest, that's why I even thought than putting into settings, since these options were the settings of it.

> I'm wondering if we can still have a freeform settings option, but have its toplevel attrs relate to keys in the `require('lint')[key] = val` assignment?

I don't think a freeform will be needed, since the plugin is quite simple, and it will probably not add anything else.